### DISCUSSION TITLE
Received message not working?

### DISCUSSION THREAD

**Discussion by GergesBernaba1** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#58] [Q&A]
I try to use the Controller received message and have WA account and sending a message and templet message and i tried to reply to test received message but not working correct i can't read the received message like it mentioned in here? 
i setup a webhook configuration but still not working, if anyone can help me i appropriate the support

**Comment by gabrieldwight** [comment]
Hi, you can refer to [Webhook subscription](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/discussions/24#discussioncomment-4317961)

If you are debugging locally you will need to use dev tunnels, conveyor or ngrok

**Reply by GergesBernaba1** [reply]
i run in production and I set up webhook and subscribe.
does this come from WhatsApp?
 public ActionResult<string> ConfigureWhatsAppMessageWebhook([FromQuery(Name = "hub.mode")] string hubMode,
                                                             [FromQuery(Name = "hub.challenge")] int hubChallenge,
                                                             [FromQuery(Name = "hub.verify_token")] string hubVerifyToken)

**Reply by gabrieldwight** [reply]
Yes it comes from whatsapp when subscribing to the endpoint that will receive messages for your phone number. If it is verified you can check if your endpoint is getting any messages from whatsapp

**Reply by GergesBernaba1** [reply]
always the error here :
        var changesResult = messageReceived["entry"][0]["changes"][0]["value"];
         An unhandled exception was thrown by the application.
Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: The best overloaded method match for 'System.Text.Json.JsonElement.this[int]' has some invalid arguments
   at CallSite.Target(Closure, CallSite, Object, String)

**Reply by gabrieldwight** [reply]
Does it happen on deserialization of the dynamic request?

**Reply by GergesBernaba1** [reply]
Yes

**Reply by gabrieldwight** [reply]
I think deserializing with system.text.json you will have to cast the dynamic object to JsonElement then you can access the individual properties since with newtonsoft does not throw the error above.

```C#

var jsonstring = JsonSerializer.Deserialize<JsonElement>(messageStr);

```

**Reply by GergesBernaba1** [reply]
thanks already used the this in Deserialize and all is well

**Reply by GergesBernaba1** [reply]
Hi i have a problem in this line and in MarkAsRead: always in Mark return missing permissions 
but in this:
var mediaFileDownloaded = await _whatsAppBusinessClient.DownloadMediaAsync(mediaUrlResponse.Url);
and when GetMediaUrlAsync
 Error when DownloadMediaAsync:  The value cannot be null or empty. (Parameter 'name')
 Error when GetMediaUrlAsync:  The value cannot be null or empty. (Parameter 'name')

**Reply by gabrieldwight** [reply]
MarkAsRead will require the message id for your whatsapp message received in the webhook. For example from the sample project when you send an audio file

```C#
if (messageType.Equals("audio"))
                    {
                        var audioMessageReceived = JsonConvert.DeserializeObject<AudioMessageReceived>(Convert.ToString(messageReceived)) as AudioMessageReceived;
                        audioMessage = new List<AudioMessage>(audioMessageReceived.Entry.SelectMany(x => x.Changes).SelectMany(x => x.Value.Messages));
                        _logger.LogInformation(JsonConvert.SerializeObject(audioMessage, Formatting.Indented));

                        MarkMessageRequest markMessageRequest = new MarkMessageRequest();
                        markMessageRequest.MessageId = audioMessage.SingleOrDefault().Id;
                        markMessageRequest.Status = "read";

                        await _whatsAppBusinessClient.MarkMessageAsReadAsync(markMessageRequest);

                        var mediaUrlResponse = await _whatsAppBusinessClient.GetMediaUrlAsync(audioMessage.SingleOrDefault().Audio.Id);

                        _logger.LogInformation(mediaUrlResponse.Url);

                        // To download media received sent by user
                        var mediaFileDownloaded = await _whatsAppBusinessClient.DownloadMediaAsync(mediaUrlResponse.Url);

                        var rootPath = Path.Combine(_webHostEnvironment.WebRootPath, "Application_Files\\MediaDownloads\\");

                        if (!Directory.Exists(rootPath))
                        {
                            Directory.CreateDirectory(rootPath);
                        }

                        // Get the path of filename
                        string filename = string.Empty;

                        if (mediaUrlResponse.MimeType.Contains("audio/mpeg"))
                        {
                            filename = $"{mediaUrlResponse.Id}.mp3";
                        }

                        if (mediaUrlResponse.MimeType.Contains("audio/ogg"))
                        {
                            filename = $"{mediaUrlResponse.Id}.ogg";
                        }

                        var filePath = Path.Combine(_webHostEnvironment.WebRootPath, "Application_Files\\MediaDownloads\\", filename);

                        await System.IO.File.WriteAllBytesAsync(filePath, mediaFileDownloaded);

                        return Ok(new
                        {
                            Message = "Audio Message received"
                        });
                    }
```
